File: src/verification/nvidia.ts
import axios from "axios";
import type { VerificationResult, Verifier } from "./types";

interface NvidiaAttestationPayload {
	nonce: string;
	arch: string;
	evidence_list: Array<{
		certificate: string;
		evidence: string;
		arch: string;
	}>;
}

export class NvidiaGpuVerifier implements Verifier {
	private readonly NRAS_URL =
		"https://nras.attestation.nvidia.com/v3/attest/gpu";

	async verify(payload: NvidiaAttestationPayload): Promise<VerificationResult> {
		try {
			// 1. Send to NRAS
			const response = await axios.post(this.NRAS_URL, payload, {
				headers: {
					accept: "application/json",
					"content-type": "application/json",
				},
			});

			if (response.status !== 200) {
				throw new Error(`NRAS responded with status ${response.status}`);
			}

			const tokens = response.data;
			if (!Array.isArray(tokens) || tokens.length < 1) {
				throw new Error("Invalid response format: expected array");
			}

			// 2. Parse the platform token (Index 0)
			// Expected format: [ ["JWT", "token..."], { "GPU-0": "token..." } ]
			// We look at the first element which should be ["JWT", "token"]
			const platformTokenEntry = tokens[0];
			if (
				!Array.isArray(platformTokenEntry) ||
				platformTokenEntry[0] !== "JWT"
			) {
				throw new Error("Invalid platform token format");
			}

			const platformJwt = platformTokenEntry[1];
			const platformClaims = this.decodeJwt(platformJwt);

			// 3. Check overall result
			const isValid = platformClaims["x-nvidia-overall-att-result"] === true;

			// 4. Extract GPU claims (from the second element if available, or from submods in platform token?)
			// The platform token has 'submods'. The second element in the array is detailed GPU tokens.
			// Let's grab the first GPU detail if available.
			let gpuClaims = {};
			if (tokens.length > 1 && typeof tokens[1] === "object") {
				const gpuTokens = tokens[1];
				const firstGpuKey = Object.keys(gpuTokens)[0]; // e.g., "GPU-0"
				if (firstGpuKey && typeof gpuTokens[firstGpuKey] === "string") {
					gpuClaims = this.decodeJwt(gpuTokens[firstGpuKey]);
				}
			}

			return {
				isValid,
				timestamp: Date.now(),
				hardwareType: "NVIDIA_CC",
				claims: {
					...platformClaims,
					...gpuClaims,
				},
				raw: tokens,
				error: isValid ? undefined : "Nvidia attestation result is false",
			};
		} catch (error: unknown) {
			const errorMessage =
				error instanceof Error ? error.message : String(error);
			return {
				isValid: false,
				timestamp: Date.now(),
				hardwareType: "NVIDIA_CC",
				claims: {},
				error: errorMessage || "Unknown error during Nvidia verification",
			};
		}
	}

	// biome-ignore lint/suspicious/noExplicitAny: JWT result is dynamic
	private decodeJwt(token: string): any {
		try {
			const parts = token.split(".");
			if (parts.length !== 3) {
				throw new Error("Invalid JWT format");
			}
			const payload = parts[1];
			if (!payload) {
				throw new Error("Payload missing");
			}
			const decoded = Buffer.from(payload, "base64").toString("utf-8");
			return JSON.parse(decoded);
		} catch (e) {
			console.error("Failed to decode JWT:", e);
			return {};
		}
	}
}

File: confidential_verifier/verifiers/nvidia.py
import logging
import requests
import json
import time
import jwt
from typing import Dict, Any
from ..types import VerificationResult, HARDWARE_NVIDIA_CC
from .base import Verifier

logger = logging.getLogger(__name__)


class NvidiaGpuVerifier(Verifier):
    def __init__(self):
        self.nras_url = "https://nras.attestation.nvidia.com/v3/attest/gpu"

    async def verify(self, payload: Dict[str, Any]) -> VerificationResult:
        try:
            response = requests.post(
                self.nras_url,
                json=payload,
                headers={
                    "accept": "application/json",
                    "content-type": "application/json",
                },
            )

            if response.status_code != 200:
                raise Exception(f"NRAS responded with status {response.status_code}")

            tokens = response.json()
            if not isinstance(tokens, list) or len(tokens) < 1:
                raise Exception("Invalid response format: expected array")

            # Platform token at index 0
            platform_entry = tokens[0]
            if not isinstance(platform_entry, list) or platform_entry[0] != "JWT":
                raise Exception("Invalid platform token format")

            platform_jwt = platform_entry[1]
            platform_claims = self._decode_jwt(platform_jwt)

            is_valid = platform_claims.get("x-nvidia-overall-att-result") is True

            gpu_claims = {}
            if len(tokens) > 1 and isinstance(tokens[1], dict):
                gpu_tokens = tokens[1]
                if gpu_tokens:
                    first_gpu_key = list(gpu_tokens.keys())[0]
                    gpu_jwt = gpu_tokens[first_gpu_key]
                    if isinstance(gpu_jwt, str):
                        gpu_claims = self._decode_jwt(gpu_jwt)

            claims = {**platform_claims, **gpu_claims}

            return VerificationResult(
                model_verified=is_valid,
                provider="nvidia",
                timestamp=time.time(),
                hardware_type=[HARDWARE_NVIDIA_CC],
                claims=claims,
                error=None if is_valid else "Nvidia attestation result is false",
            )

        except Exception as e:
            logger.exception("Nvidia verification failed")
            return VerificationResult(
                model_verified=False,
                provider="nvidia",
                timestamp=time.time(),
                hardware_type=[],
                claims={},
                error=str(e),
            )

    def _decode_jwt(self, token: str) -> Dict[str, Any]:
        try:
            # TODO: Fetch JWKS from Nvidia and enable signature verification for better security.
            return jwt.decode(
                token,
                options={"verify_signature": False},
                algorithms=["RS256", "ES256", "ES384", "PS256"],
            )
        except Exception as e:
            logger.error(f"Failed to decode JWT: {e}")
            return {}
